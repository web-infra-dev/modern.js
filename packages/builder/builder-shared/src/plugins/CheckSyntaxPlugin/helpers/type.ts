import { CheckSyntaxOptions } from '../../../types';

export type CheckSyntaxExclude = NonNullable<CheckSyntaxOptions['exclude']>;

export interface Location {
  line: number;
  column: number;
}

export interface File {
  path: string;
  line: number;
  column: number;
}

interface SyntaxErrorOptions {
  source: File & { code: string };
  output?: File;
}

export class SyntaxError extends Error {
  source: File & { code: string };

  output: File | undefined;

  constructor(message: string, options: SyntaxErrorOptions) {
    super(message);
    this.source = options.source;
    this.output = options.output;
  }
}

export type AcornParseError = {
  message: string;
  pos: number;
  loc: Location;
};
