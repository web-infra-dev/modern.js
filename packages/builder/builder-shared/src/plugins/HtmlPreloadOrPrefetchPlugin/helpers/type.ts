import type HtmlWebpackPlugin from 'html-webpack-plugin';

export type BeforeAssetTagGenerationHtmlPluginData = {
  assets: {
    publicPath: string;
    js: Array<string>;
    css: Array<string>;
    favicon?: string;
    manifest?: string;
  };
  outputName: string;
  plugin: HtmlWebpackPlugin;
};

export type As =
  | 'audio'
  | 'document'
  | 'embed'
  | 'fetch'
  | 'font'
  | 'image'
  | 'object'
  | 'script'
  | 'style'
  | 'track'
  | 'worker'
  | 'video';
