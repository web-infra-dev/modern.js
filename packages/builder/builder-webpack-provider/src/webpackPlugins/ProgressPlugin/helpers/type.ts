import type { ForegroundColor as Color } from '@modern-js/utils/compiled/chalk';

export type Props = {
  total: number;
  current: number;
  color: typeof Color;
  bgColor: typeof Color;
  char: string;
  width: number;
  buildIcon: string;
  finishIcon: string;
  finishInfo: string;
  message: string;
  done: boolean;
  messageWidth: number;
  spaceWidth: number;
  messageColor: typeof Color;
  id: string;
  maxIdLen: number;
};

export type BusOption = {
  state: Props[];
};
