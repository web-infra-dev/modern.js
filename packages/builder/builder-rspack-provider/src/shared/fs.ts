import { join } from 'path';

export const getCompiledPath = (packageName: string) =>
  join(__dirname, '../../compiled', packageName);
