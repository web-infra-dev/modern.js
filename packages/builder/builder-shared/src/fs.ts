import { isAbsolute, join } from 'path';
import { MODULE_PATH_REGEX } from './constants';
import { removeLeadingSlash } from './utils';
import {
  DistPathConfig,
  NormalizedSharedOutputConfig,
  SharedHtmlConfig,
  FilenameConfig,
} from './types';

export function getAbsoluteDistPath(
  cwd: string,
  outputConfig: NormalizedSharedOutputConfig,
) {
  const root = getDistPath(outputConfig, 'root');
  return isAbsolute(root) ? root : join(cwd, root);
}

export const getDistPath = (
  outputConfig: NormalizedSharedOutputConfig,
  type: keyof DistPathConfig,
): string => {
  const { distPath } = outputConfig;
  const ret = distPath[type];
  if (typeof ret !== 'string') {
    throw new Error(`unknown key ${type} in "output.distPath"`);
  }
  return ret;
};

export async function isFileExists(file: string) {
  const { promises, constants } = await import('fs');
  return promises
    .access(file, constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export function getPackageNameFromModulePath(modulePath: string) {
  const handleModuleContext = modulePath?.match(MODULE_PATH_REGEX);

  if (!handleModuleContext) {
    return false;
  }

  const [, , scope, name] = handleModuleContext;
  const packageName = ['npm', (scope ?? '').replace('@', ''), name]
    .filter(Boolean)
    .join('.');

  return packageName;
}

export function getHTMLPathByEntry(
  entryName: string,
  config: {
    output: NormalizedSharedOutputConfig;
    html: SharedHtmlConfig;
  },
) {
  const htmlPath = getDistPath(config.output, 'html');
  const filename = config.html.disableHtmlFolder
    ? `${entryName}.html`
    : `${entryName}/index.html`;

  return removeLeadingSlash(`${htmlPath}/${filename}`);
}

export const getFilename = (
  output: NormalizedSharedOutputConfig,
  type: keyof FilenameConfig,
  isProd: boolean,
) => {
  const { filename } = output;
  const useHash = !output.disableFilenameHash;
  const hash = useHash ? '.[contenthash:8]' : '';

  switch (type) {
    case 'js':
      return filename.js ?? `[name]${isProd ? hash : ''}.js`;
    case 'css':
      return filename.css ?? `[name]${isProd ? hash : ''}.css`;
    case 'svg':
      return filename.svg ?? `[name]${hash}.svg`;
    case 'font':
      return filename.font ?? `[name]${hash}[ext]`;
    case 'image':
      return filename.image ?? `[name]${hash}[ext]`;
    case 'media':
      return filename.media ?? `[name]${hash}[ext]`;
    default:
      throw new Error(`unknown key ${type} in "output.filename"`);
  }
};
