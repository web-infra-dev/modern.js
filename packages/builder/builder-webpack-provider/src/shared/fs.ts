import { join } from 'path';
import {
  CSS_DIST_DIR,
  FONT_DIST_DIR,
  HTML_DIST_DIR,
  IMAGE_DIST_DIR,
  JS_DIST_DIR,
  MEDIA_DIST_DIR,
  ROOT_DIST_DIR,
  SVG_DIST_DIR,
} from '@modern-js/builder-shared';
import type { BuilderConfig, DistPathConfig, FilenameConfig } from '../types';

export const getCompiledPath = (packageName: string) =>
  join(__dirname, '../../compiled', packageName);

export const getDistPath = (
  config: BuilderConfig,
  type: keyof DistPathConfig,
): string => {
  const { distPath = {} } = config.output || {};

  switch (type) {
    case 'js':
      return distPath.js ?? JS_DIST_DIR;
    case 'css':
      return distPath.css ?? CSS_DIST_DIR;
    case 'svg':
      return distPath.svg ?? SVG_DIST_DIR;
    case 'font':
      return distPath.font ?? FONT_DIST_DIR;
    case 'html':
      return distPath.html ?? HTML_DIST_DIR;
    case 'media':
      return distPath.media ?? MEDIA_DIST_DIR;
    case 'root':
      return distPath.root ?? ROOT_DIST_DIR;
    case 'image':
      return distPath.image ?? IMAGE_DIST_DIR;
    default:
      throw new Error(`unknown key ${type} in "output.distPath"`);
  }
};

export const getFilename = (
  config: BuilderConfig,
  type: keyof FilenameConfig,
  isProd: boolean,
) => {
  const { filename = {} } = config.output || {};
  const useHash = isProd && !config.output?.disableFilenameHash;
  const hash = useHash ? '.[contenthash:8]' : '';

  switch (type) {
    case 'js':
      return filename.js ?? `[name]${hash}.js`;
    case 'css':
      return filename.css ?? `[name]${hash}.css`;
    case 'svg':
      return filename.svg ?? `[name]${hash}.svg`;
    case 'font':
      return filename.font ?? `[name]${hash}[ext]`;
    case 'image':
      return filename.image ?? `[name]${hash}[ext]`;
    case 'media':
      return filename.media ?? `[name]${hash}[ext]`;
    default:
      throw new Error(`unknown key ${type} in "output.filename"`);
  }
};
