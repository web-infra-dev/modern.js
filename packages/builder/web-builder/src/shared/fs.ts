import { join } from 'path';
import type { BuilderConfig, DistPathConfig } from '../types';
import {
  JS_DIST_DIR,
  CSS_DIST_DIR,
  SVG_DIST_DIR,
  FONT_DIST_DIR,
  ROOT_DIST_DIR,
  HTML_DIST_DIR,
  IMAGE_DIST_DIR,
  MEDIA_DIST_DIR,
} from './constants';

export async function isFileExists(file: string) {
  const { promises, constants } = await import('fs');
  return promises
    .access(file, constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export const getCompiledPath = (packageName: string) =>
  join(__dirname, '../../compiled', packageName);

export const getDistPath = (
  config: BuilderConfig,
  type: keyof DistPathConfig,
) => {
  const { distPath } = config.output || {};
  const distConfig =
    typeof distPath === 'object' ? distPath : { root: distPath };

  switch (type) {
    case 'js':
      return distConfig.js || JS_DIST_DIR;
    case 'css':
      return distConfig.css || CSS_DIST_DIR;
    case 'svg':
      return distConfig.svg || SVG_DIST_DIR;
    case 'font':
      return distConfig.font || FONT_DIST_DIR;
    case 'html':
      return distConfig.html || HTML_DIST_DIR;
    case 'media':
      return distConfig.media || MEDIA_DIST_DIR;
    case 'root':
      return distConfig.root || ROOT_DIST_DIR;
    case 'image':
      return distConfig.image || IMAGE_DIST_DIR;
    default:
      throw new Error('unknown type');
  }
};
