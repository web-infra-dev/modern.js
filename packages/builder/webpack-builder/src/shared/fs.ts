import { join } from 'path';
import type { BuilderConfig, DistPathConfig, FilenameConfig } from '../types';

export async function isFileExists(file: string) {
  const { promises, constants } = await import('fs');
  return promises
    .access(file, constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export const getCompiledPath = (packageName: string) =>
  join(__dirname, '../../compiled', packageName);

export const getDistPath = (
  config: BuilderConfig,
  type: keyof DistPathConfig,
) => {
  const { distPath = {} } = config.output;

  switch (type) {
    case 'js':
      return distPath.js;
    case 'css':
      return distPath.css;
    case 'svg':
      return distPath.svg;
    case 'font':
      return distPath.font;
    case 'html':
      return distPath.html;
    case 'media':
      return distPath.media;
    case 'root':
      return distPath.root;
    case 'image':
      return distPath.image;
    default:
      throw new Error(`unknown key ${type} in "output.distPath"`);
  }
};

export const getFilename = (
  config: BuilderConfig,
  type: keyof FilenameConfig,
  isProd: boolean,
) => {
  const { filename = {} } = config.output || {};
  const useHash = isProd && !config.output?.disableFilenameHash;
  const hash = useHash ? '.[contenthash:8]' : '';

  switch (type) {
    case 'js':
      return filename.js ?? `[name]${hash}.js`;
    case 'css':
      return filename.css ?? `[name]${hash}.css`;
    case 'svg':
      return filename.svg ?? `[name]${hash}.svg`;
    case 'font':
      return filename.font ?? `[name]${hash}[ext]`;
    case 'image':
      return filename.image ?? `[name]${hash}[ext]`;
    case 'media':
      return filename.media ?? `[name]${hash}[ext]`;
    default:
      throw new Error(`unknown key ${type} in "output.filename"`);
  }
};
