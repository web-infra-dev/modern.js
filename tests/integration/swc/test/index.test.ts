import path from 'path';
import { readdirSync, readFileSync } from 'fs';
import type { ChildProcess } from 'child_process';
import {
  modernBuild,
  runModernCommandDev,
} from '../../../utils/modernTestUtils';

const fixtures = path.resolve(__dirname, '../fixtures');

const getJsFiles = (appDir: string) =>
  readdirSync(path.resolve(appDir, 'dist/static/js')).filter(filepath =>
    /\.js$/.test(filepath),
  );

const getCssFiles = (appDir: string) =>
  readdirSync(path.resolve(appDir, 'dist/static/css')).filter(filepath =>
    /\.css$/.test(filepath),
  );

describe('swc minify', () => {
  it('should emitted script files correctly', async () => {
    const appDir = path.resolve(fixtures, 'minify-js');

    await modernBuild(appDir);

    const files = getJsFiles(appDir);
    const mainFile = files.filter(filepath => filepath.startsWith('main'))[0];

    expect(files.length).toBe(3);
    expect(
      readFileSync(path.resolve(appDir, `dist/static/js/${mainFile}`), 'utf8'),
    ).toContain('children:"helloworld"');
  });

  it('should not exit unexpectly when transform failed', async () => {
    const appDir = path.resolve(fixtures, 'transform-fail');

    const cp: ChildProcess = await runModernCommandDev(['dev'], false, {
      cwd: appDir,
      env: {
        NODE_ENV: 'development',
      },
    });

    expect(cp).toBeDefined();
    expect(cp.exitCode).toBe(null);

    cp.kill('SIGTERM');
  });

  it('should minify css', async () => {
    const appDir = path.resolve(fixtures, 'minify-css');

    await modernBuild(appDir);

    const cssFile = getCssFiles(appDir)[0];
    expect(cssFile).toMatchSnapshot('swc minified css');
  });
});
