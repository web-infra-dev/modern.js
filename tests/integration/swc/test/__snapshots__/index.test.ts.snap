// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swc minify should minify css: swc minified css 1`] = `"main.fa69b7c4.css"`;
